### PR TITLE
[SPARK-46272][SQL] Support CTAS using DSv2 sources

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

https://github.com/apache/spark/pull/43949 supports CREATE TABLE using DSv2 sources. This PR supports CREATE TABLE AS SELECT (CTAS) using DSv2 sources. It turns out that we don't need additional code changes. This PR simply adds more test cases for CTAS queries.

### Why are the changes needed?

To add tests for CTAS for DSv2 sources.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

New tests

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 1]
```diff
@@ -914,8 +914,8 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase with Compilat
     unsupportedTableOperationError(table.name(), "either micro-batch or continuous scan")
   }
 
-  def unsupportedAppendInBatchModeError(table: Table): Throwable = {
-    unsupportedTableOperationError(table.name(), "append in batch mode")
+  def unsupportedAppendInBatchModeError(name: String): Throwable = {
+    unsupportedTableOperationError(name, "append in batch mode")
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:918]: When we change `src/main`, you don't need to use `[TESTS]`, @allisonwang-db .

Comment (allisonwang-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:918]: I see! Thanks for letting me know!

**Comment by allisonwang-db** [comment]
cc @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala [hunk 2]
```diff
@@ -723,6 +724,158 @@ class DataSourceV2Suite extends QueryTest with SharedSparkSession with AdaptiveS
...
+               |CREATE TABLE test USING ${cls.getName}
+               |AS VALUES ('a', 'b'), ('c', 'd')
+               |""".stripMargin)
+        },
+        errorClass = "INCOMPATIBLE_DATA_FOR_TABLE.CANNOT_SAFELY_CAST",
+        parameters = Map(
+          "tableName" -> "`spark_catalog`.`default`.`test`",
+          "colName" -> "`i`", "srcType" -> "\"STRING\"", "targetType" -> "\"INT\""))
+    }
+    // CTAS with schema specified
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala:801]: This is not related to data source and we probably don't need it here.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala [hunk 3]
```diff
@@ -723,6 +724,158 @@ class DataSourceV2Suite extends QueryTest with SharedSparkSession with AdaptiveS
...
+    withTable("test") {
+      checkError(
+        exception = intercept[AnalysisException] {
+          sql(
+            s"""
+               |CREATE TABLE test USING ${cls.getName}
+               |AS VALUES (0, 1, 2), (1, 2, 3)
+               |""".stripMargin)
+        },
+        errorClass = "INSERT_COLUMN_ARITY_MISMATCH.TOO_MANY_DATA_COLUMNS",
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala:764]: I'm a bit confused here. So for this data source, we probably don't check anything when creating the table. But when load the table, and get the `Table` instance in `V2SessionCatalog`, shall we check if `Table#schema` match what we stored in HMS?

Comment (allisonwang-db) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala:764]: This is actually tricky, and behavior depends on the implementation of the data source.

In `createTable`, we store the query's schema (col1, col2, col3) in HMS, but this data source's `getTable` method does not take into account the input `schema`.

So, when the DSv2 relation is being created, it uses `table.columns.asSchema` which is the default implementation of the schema (i, j).  
https://github.com/apache/spark/blob/bacdb3b5fec9783f46042764eeee80eb2a0f5702/sql/catalyst/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala#L210-L211

However, if the data source utilize user-defined schema in `getTable`, it won't throw this exception.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala:758]: I think we should enforce something here. When calling `TableProvider.getTable` with user-specified schema, we should make sure the returned table reports the same schema as the user-specified schema (probably ignore nullability). It's actually a documented requirement
```
   * Return a {@link Table} instance with the specified table schema, partitioning and properties
   * to do read/write. The returned table should report the same schema and partitioning with the
   * specified ones, or Spark may fail the operation.
```

File: common/utils/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -887,6 +887,11 @@
     ],
     "sqlState" : "42K02"
   },
+  "DATA_SOURCE_TABLE_SCHEMA_MISMATCH" : {
+    "message" : [
+      "The schema of the data source table <tableSchema> does not match the actual schema <actualSchema>. If you are using the DataFrameReader.schema API or creating a table, avoid specifying the schema."
+    ]
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-classes.json:893]: sqlState is required now. maybe `42K02` is fine

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala [hunk 5]
```diff
@@ -85,6 +86,11 @@ case class CreateTableAsSelectExec(
     val table = catalog.createTable(
       ident, getV2Columns(query.schema, catalog.useNullableQuerySchema),
       partitioning.toArray, properties.asJava)
+    // Check if the table schema matches the schema of the query.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala:89]: shall we do this check in a lower level? where we instantiate the v2 Table.

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala [hunk 6]
```diff
@@ -725,8 +725,127 @@ class DataSourceV2Suite extends QueryTest with SharedSparkSession with AdaptiveS
...
+      sql(
+        s"""
+           |CREATE TABLE IF NOT EXISTS test USING ${cls.getName}
+           |AS VALUES (3, 4), (4, 5)
+           |""".stripMargin)
+      checkAnswer(sql("SELECT * FROM test"), Seq((2, 3), (4, 5)).toDF("i", "j"))
+    }
+  }
+
+  test("SPARK-46272: create table as select - schema name mismatch") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala:752]: can we have create table test (no as select) for schema mismatch?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala (hunk 5 above)
Comment (allisonwang-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/WriteToDataSourceV2Exec.scala:89]: Make sense

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala [hunk 7]
```diff
@@ -232,7 +233,19 @@ class V2SessionCatalog(catalog: SessionCatalog)
...
     }
 
-    loadTable(ident)
+    val table = loadTable(ident)
+
+    // Check if the schema of the created table matches the given schema.
+    if (schema.nonEmpty) {
+      val tableSchema = CharVarcharUtils.replaceCharVarcharWithStringInSchema(
+        table.columns().asSchema)
+      if (!DataType.equalsIgnoreNullability(tableSchema, schema)) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala:242]: can we check it in `loadTable`? In case the data source is a bit random and only return wrong schema at second time.

Comment (allisonwang-db) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2SessionCatalog.scala:242]: I tried it but looks like `loadTable` is used in many other places, such as in `tableExists`. If we check the schema there, it will fail commands other than create table (e.g `DROP TABLE t` will fail because the schema in catalogTable does not match the table schema)

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
LGTM if all tests pass

**Review by cloud-fan — APPROVED** [review]

**Comment by allisonwang-db** [comment]
@cloud-fan the test failure seems unrelated

**Comment by cloud-fan** [comment]
Can you retrigger the GC jobs?

**Comment by cloud-fan** [comment]
thanks, merging to master!
